File: benchmarks/go/compiler_ast.go
package main

import "fmt"

type Expr interface{}

type ELit struct {
	value int
}

type EVar struct {
	id int
}

type EAdd struct {
	left  Expr
	right Expr
}

type EMul struct {
	left  Expr
	right Expr
}

type ELet struct {
	name  int
	value Expr
	body  Expr
}

type EIf struct {
	cond Expr
	then Expr
	els  Expr
}

func buildExpr(depth int, seed int) Expr {
	if depth <= 0 {
		if seed%3 == 0 {
			return EVar{seed % 4096}
		}
		return ELit{(seed*17 + 11) % 100000}
	}
	next := seed*3 + depth
	switch depth % 5 {
	case 0:
		return EAdd{buildExpr(depth-1, next), buildExpr(depth-1, next+1)}
	case 1:
		return EMul{buildExpr(depth-1, next), buildExpr(depth-1, next+3)}
	case 2:
		return ELet{seed % 8192, buildExpr(depth-1, next), buildExpr(depth-1, next+5)}
	case 3:
		return EIf{
			buildExpr(depth-1, next),
			buildExpr(depth-2, next+7),
			buildExpr(depth-2, next+11),
		}
	default:
		return EAdd{buildExpr(depth-1, next+13), buildExpr(depth-1, next+17)}
	}
}

func rewriteExpr(expr Expr, pass int) Expr {
	switch e := expr.(type) {
	case ELit:
		return ELit{(e.value + pass*13 + 7) % 1000003}
	case EVar:
		return EVar{(e.id*33 + pass + 19) % 16384}
	case EAdd:
		if pass%7 == 0 {
			return EAdd{rewriteExpr(e.right, pass), rewriteExpr(e.left, pass)}
		}
		return EAdd{rewriteExpr(e.left, pass), rewriteExpr(e.right, pass)}
	case EMul:
		return EMul{rewriteExpr(e.left, pass), rewriteExpr(e.right, pass)}
	case ELet:
		return ELet{(e.name + pass) % 16384, rewriteExpr(e.value, pass), rewriteExpr(e.body, pass+1)}
	case EIf:
		return EIf{
			rewriteExpr(e.cond, pass),
			rewriteExpr(e.then, pass+2),
			rewriteExpr(e.els, pass+3),
		}
	default:
		return ELit{0}
	}
}

func checksumExpr(expr Expr) int {
	switch e := expr.(type) {
	case ELit:
		return e.value + 3
	case EVar:
		return e.id*5 + 7
	case EAdd:
		return 11 + checksumExpr(e.left)*3 + checksumExpr(e.right)
	case EMul:
		return 17 + checksumExpr(e.left) + checksumExpr(e.right)*3
	case ELet:
		return 23 + e.name + checksumExpr(e.value) + checksumExpr(e.body)*5
	case EIf:
		return 31 + checksumExpr(e.cond) + checksumExpr(e.then)*7 + checksumExpr(e.els)
	default:
		return 0
	}
}

func runPipeline(depth int, passes int) int {
	expr := buildExpr(depth, 19)
	checksum := 0
	for pass := 0; pass < passes; pass++ {
		expr = rewriteExpr(expr, pass)
		if pass%8 == 0 {
			checksum += checksumExpr(expr)
		}
	}
	return checksum + checksumExpr(expr)
}

func main() {
	checksum := runPipeline(11, 80)
	fmt.Printf("compiler_ast checksum: %d\n", checksum)
}

File: benchmarks/ocaml/compiler_ast.ml
(* Compiler-shaped AST pass benchmark.
   Models repeated recursive IR rewrites over immutable tree nodes. *)

type expr =
  | ELit of int
  | EVar of int
  | EAdd of expr * expr
  | EMul of expr * expr
  | ELet of int * expr * expr
  | EIf of expr * expr * expr

let rec build_expr depth seed =
  if depth <= 0 then
    if seed mod 3 = 0 then EVar (seed mod 4096)
    else ELit ((seed * 17 + 11) mod 100000)
  else
    let next = (seed * 3) + depth in
    match depth mod 5 with
    | 0 -> EAdd (build_expr (depth - 1) next, build_expr (depth - 1) (next + 1))
    | 1 -> EMul (build_expr (depth - 1) next, build_expr (depth - 1) (next + 3))
    | 2 -> ELet (seed mod 8192, build_expr (depth - 1) next, build_expr (depth - 1) (next + 5))
    | 3 ->
        EIf
          ( build_expr (depth - 1) next,
            build_expr (depth - 2) (next + 7),
            build_expr (depth - 2) (next + 11) )
    | _ -> EAdd (build_expr (depth - 1) (next + 13), build_expr (depth - 1) (next + 17))

let rec rewrite_expr expr pass =
  match expr with
  | ELit value -> ELit ((value + (pass * 13) + 7) mod 1000003)
  | EVar id -> EVar (((id * 33) + pass + 19) mod 16384)
  | EAdd (left, right) ->
      if pass mod 7 = 0 then EAdd (rewrite_expr right pass, rewrite_expr left pass)
      else EAdd (rewrite_expr left pass, rewrite_expr right pass)
  | EMul (left, right) -> EMul (rewrite_expr left pass, rewrite_expr right pass)
  | ELet (name, value, body) ->
      ELet ((name + pass) mod 16384, rewrite_expr value pass, rewrite_expr body (pass + 1))
  | EIf (cond, then_expr, else_expr) ->
      EIf
        ( rewrite_expr cond pass,
          rewrite_expr then_expr (pass + 2),
          rewrite_expr else_expr (pass + 3) )

let rec checksum_expr expr =
  match expr with
  | ELit value -> value + 3
  | EVar id -> (id * 5) + 7
  | EAdd (left, right) -> 11 + (checksum_expr left * 3) + checksum_expr right
  | EMul (left, right) -> 17 + checksum_expr left + (checksum_expr right * 3)
  | ELet (name, value, body) -> 23 + name + checksum_expr value + (checksum_expr body * 5)
  | EIf (cond, then_expr, else_expr) ->
      31 + checksum_expr cond + (checksum_expr then_expr * 7) + checksum_expr else_expr

let run_pipeline depth passes =
  let expr = ref (build_expr depth 19) in
  let checksum = ref 0 in
  for pass = 0 to passes - 1 do
    expr := rewrite_expr !expr pass;
    if pass mod 8 = 0 then checksum := !checksum + checksum_expr !expr
  done;
  !checksum + checksum_expr !expr

let () =
  let checksum = run_pipeline 11 80 in
  Printf.printf "compiler_ast checksum: %d\n" checksum
